src/Workspaces/Core/MSBuild/MSBuild/MSBuildProjectLoader.Worker_ResolveReferences.cs

```diff
@@ -0,0 +1,291 @@
+                                        done = true;
+                                        break;
+                                    }
+                                }
+                            }
+
+                            if (done)
+                            {
+                                continue;
+                            }
+
+                            // We didn't find the output file path in this project's metadata references, or it doesn't exist on disk.
+                            // In that case, carry on and load the project.
+                        }
+
+                        // OK, we've got to load the project.
+                        var projectRefInfos = await LoadProjectInfosFromPathAsync(absoluteProjectPath, _discoveredProjectOptions, cancellationToken).ConfigureAwait(false);
+
+                        var projectRefInfo = metadataReferenceSet.Find(projectRefInfos);
+                        if (projectRefInfo != null)
+                        {
+                            // Don't add a reference if the project already has a reference on us. Otherwise, it will cause a circularity.
+                            if (projectRefInfo.ProjectReferences.Any(pr => pr.ProjectId == id))
+                            {
+                                // If the metadata doesn't exist on disk, we'll have to remove it from the metadata references.
+
+                                if (!File.Exists(projectRefInfo.OutputRefFilePath))
+                                {
+                                    metadataReferenceSet.Remove(projectRefInfo.OutputRefFilePath);
+                                }
+
+                                if (!File.Exists(projectRefInfo.OutputFilePath))
+                                {
+                                    metadataReferenceSet.Remove(projectRefInfo.OutputFilePath);
+                                }
+                            }
+                            else
+                            {
+                                projectReferences.Add(CreateProjectReference(id, projectRefInfo.Id, projectFileReference.Aliases));
+                                metadataReferenceSet.Remove(projectRefInfo.OutputRefFilePath);
+                                metadataReferenceSet.Remove(projectRefInfo.OutputFilePath);
+                            }
+
+                            continue;
+                        }
+                    }
+
+                    // We weren't able to handle this project reference, so add it without further processing.
+                    var unknownProjectId = _projectMap.GetOrCreateProjectId(projectFileReference.Path);
+                    projectReferences.Add(CreateProjectReference(id, unknownProjectId, projectFileReference.Aliases));
+                }
+
+                return new ResolvedReferences(
+                    projectReferences.ToImmutableArray(),
+                    metadataReferenceSet.GetRemaining());
```

Comment:
>metadataReferenceSet.GetRemaining() [](start = 20, length = 35)

Possible source of non-deterministic behavior here. I'm unsure of the behavior of this method but given it's a set I assume this is unordered.

Reply:
Yeah, this is tricky. The goal of this method is to convert metadata references passed to the compiler to project references by matching the metadata reference path with the `/ref-out` or `/out` of the project. That is, unless the user specified that they *want* to use metadata references rather than loading additional projects. However, even in that case, we can only honor the user's request for metadata references that exist on disk (since there could be projects involved that aren't built). Whew!

`metadataReferenceSet` preserves the original order of metadata references passed to the compiler. It's a set, but it's an ordered set. `projectReferences` should be in the order that the project references appeared in the user's project. However, because there's a dependency on something existing on disk, this isn't realistically deterministic.